interfaces/builtin/mount_control.go

```diff
@@ -0,0 +1,337 @@
+	"nodev",
+	"nodiratime",
+	"noexec",
+	"nolazytime",
+	"norelatime",
+	"nosuid",
+	"nostrictatime",
+	"nouser",
+	"relatime",
+	"strictatime",
+	"sync",
+	"ro",
+	"rw",
+}
+
+// mountControlInterface allows creating transient and persistent mounts
+type mountControlInterface struct {
+	commonInterface
+}
+
+// TODO: review the regexp
+// We prohibit expressions containing a comma, as that is used by AppArmor to
+// separate rules; otherwise, one could specify a path like
+//   $SNAP/foo, /** rw
+// and get read-write access to the whole filesystem.
+var whereRegexp = regexp.MustCompile(`^(/media|\$SNAP|\$SNAP_COMMON|\$SNAP_DATA)/[^,]*$`)
+
+// Excluding spaces and other characters which might allow constructing a
+// malicious string like
+//   auto) options=() /malicious/content /var/lib/snapd/hostfs/...,\n mount fstype=(
+var typeRegexp = regexp.MustCompile(`^[a-z0-9-]+$`)
+
+type MountInfo struct {
+	what       string
+	where      string
+	persistent bool
+	types      []string
+	options    []string
+}
+
+func parseStringList(mountEntry map[string]interface{}, fieldName string) ([]string, error) {
+	var list []string
+	value, ok := mountEntry[fieldName]
+	if ok {
+		interfaceList, ok := value.([]interface{})
+		if !ok {
+			return nil, fmt.Errorf(`mount-control "mount.%s" must be an array of strings (got %q)`, fieldName, value)
+		}
+		for _, iface := range interfaceList {
+			valueString, ok := iface.(string)
+			if !ok {
+				return nil, fmt.Errorf(`mount-control "mount.%s" element not a string (%q)`, fieldName, iface)
+			}
+			list = append(list, valueString)
+		}
+	}
+	return list, nil
+}
+
+func enumerateMounts(plug interfaces.Attrer, fn func(mountInfo *MountInfo) error) error {
+	mountAttr, ok := plug.Lookup("mount")
+	if !ok {
+		return nil
+	}
+	mounts, ok := mountAttr.([]interface{})
+	if !ok {
+		return errors.New(mountAttrNotADictionary)
+	}
+
+	for _, m := range mounts {
+		mount, ok := m.(map[string]interface{})
+		if !ok {
+			return errors.New(mountAttrNotADictionary)
+		}
+
+		what, ok := mount["what"].(string)
+		if !ok {
+			return fmt.Errorf(`mount-control "mount.what" must be a string`)
+		}
+
+		where, ok := mount["where"].(string)
+		if !ok {
+			return fmt.Errorf(`mount-control "mount.where" must be a string`)
+		}
+
+		persistent := false
+		persistentValue, ok := mount["persistent"]
+		if ok {
+			if persistent, ok = persistentValue.(bool); !ok {
+				return fmt.Errorf(`mount-control "mount.persistent" must be a boolean`)
+			}
+		}
+
+		types, err := parseStringList(mount, "type")
+		if err != nil {
+			return err
+		}
+
+		options, err := parseStringList(mount, "options")
+		if err != nil {
+			return err
+		}
+
+		mountInfo := &MountInfo{
+			what:       what,
+			where:      where,
+			persistent: persistent,
+			types:      types,
+			options:    options,
+		}
+
+		if err := fn(mountInfo); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func validateWhatAttr(what string) error {
+	if what[0] != '/' {
```

Comment:
I'm a bit torn on this function, on the one hand, this interface is super-privileged so nothing will be uploadable to the store to be released without approval, but on the other hand, we've seen many instances of using `system-files` where a user declares something super wildly permissive and finds that it "just works" locally during development and goes to upload only to find that we actually don't encourage/allow/condone/permit that sort of usage of `system-files`, and many days I find myself wishing that whomever used system-files this way would have been warned during development that what they are doing is not encouraged or in many ways supported.

So on the one hand, I'm looking ahead and thinking of how people could abuse this when all they need to have to make this work locally is to have an absolute path that is clean, but on the other hand there may be legitimate use cases for mounting /sys/fs/kernel/give/me/all/your/secrets to /snap/foo/current/not-your-secrets.txt and we don't want to have to push out an update to snapd to allow this at the last minute like we have had to for other interfaces.

Also, I think system-files is uniquely more likely to be abused since it is about the single most frustrating aspect of apparmor confinement for snaps - arbitrary file access, whereas this interface is about arbitrary mount access which is not nearly as common of a requirement as arbitrary file access.